### DISCUSSION TITLE
Simulating the iCub hand and arm

### DISCUSSION THREAD

**Discussion by teresalves** [robotology/community#366] [Q&A] [answered]
Hello,

For my Master Thesis I will be using the iCub hand and arm to grasp objects. I was wondering how I should go about that. The iCub simulator does not simulate the hands, correct? Is there anything in Gazebo I can use? Or what simulator is best?

Thank you,

Teresa Alves

**Comment by pattacini** [comment, marked as answer]
Hello @teresalves 

Let me give you a brief overview of the software components you may consider for your thesis work.

For controlling the arm to implement reaching, we have a quite solid and well-established layer that is responsible for doing Inverse Kinematics and trajectory generation. You can browse all the relevant resources at http://www.icub.org/doc/icub-main/icub_cartesian_interface.html. Moreover, [gazing][1] is also available. This library is abstract enough to be used in conjunction with whatever simulator we have in our portfolio.

Instead, to judge how well the iCub hand can grasp objects is a matter of which specific simulation environment is used, as per the following considerations. 

### iCub_SIM
Our traditional simulator is still [iCub_SIM][2], which can deal with the hands, although it is not properly up to the physics of manipulation. This is essentially due to two reasons:
- the hand itself is not reproduced in all the required details;
- the underlying physical engine does not perform sufficiently well in this respect.

However, you could still employ iCub_SIM depending on the level of accuracy you aim to attain, considering that you can exploit a bunch of tricks to get around the lack of performance in dealing with collisions; for example, you might want to rely on the [magnet function][3]. Of course, this would be valid as long as the tricks won't affect the core part of your study.

Find below a couple of nice results that built on iCub_SIM to explore the interaction of the hand with the environment:

1. Here's the blog maintained by **Lukáš Zdechovan** on the grasping modeled with neural networks: http://masterthesis.zdechovan.com.

https://user-images.githubusercontent.com/3738070/103001758-e7d0d580-452d-11eb-8c3f-294b702ffcda.mp4

2. The paper ["Catching Objects in Flight"][4] from **Aude Billard**'s group at LASA, whose video is also available on the web:

https://user-images.githubusercontent.com/3738070/103001778-ef907a00-452d-11eb-8172-8687d7c14cb9.mp4

### iCub Simulator in Gazebo
Some time ago we decided to phase out iCub_SIM and switch to Gazebo for a number of reasons including the fact that Gazebo has become somehow the reference in the robotics community.

You can find instructions on how to use Gazebo within our ecosystem at https://github.com/robotology/gazebo-yarp-plugins.

Although we've obtained great achievements in recent years with Gazebo, especially in the field of dynamic whole-body control (watch the results posted at https://dic.iit.it), we cannot consider the transition completed for what concerns manipulation as well as the porting of the devices we require (e.g. a proper implementation of the skin is still missing).

In particular, all the set of routines that we use for doing grasping (e.g. gazing, reaching, control of the fingers) still need to be properly tuned up. Anyway, @xEnVrE has made recently a huge leap in this direction, providing the following improvements I can point you at if you'd like to give them a shot:
- https://github.com/robotology-playground/visual-tactile-localization-simulation/tree/devel makes available the `.world` with the table and one object to play with, plus some insights on how to add further objects and a good `yarpmanager` application to start up the components.
- https://github.com/xEnVrE/gazebo-yarp-plugins/tree/visual_tactile_loc_plugins_yarp3 lets gazing work within Gazebo.
- https://github.com/xEnVrE/icub-gazebo/commits/visual_tactile_loc fiddles with the dynamical properties of the fingers to prevent them from moving crazily, plus an RSDF file, which is a scriptable SDF where you can set up friction coefficients that are central to the grasping setting.

Quite a big chunk of work that allowed us eventually to get this interesting outcome:

![issue-qa](https://user-images.githubusercontent.com/3738070/67617730-79d43580-f7e6-11e9-86a3-7b5c1d1effbe.gif)

It will be certainly our job in the upcoming quarters to port @xEnVrE patches into the mainstream.

### Alternatives
Lately, the community in Reinforcement Learning has been focusing on the use of different simulators as [V-REP][5] and [PyBullet][6] that seem very promising too. We're making progress also on this:
- [`pybullet-robot-envs`][7] is a Python package to develop and test RL algorithms for grasping (@giuliavezzani @eleramp). It's still in Alpha and doesn't have a complete hand yet.
- [`gym-ignition`][8] creates reproducible robotics environments for research in RL combining OpenAI Gym capabilities within Gazebo (@diegoferigo).

[1]: http://www.icub.org/doc/icub-main/icub_gaze_interface.html
[2]: https://wiki.icub.org/wiki/Simulator_README 
[3]: https://wiki.icub.org/wiki/Simulator_README#GRAB_FUNCTION_.28MAGNET.29:
[4]: http://lasa.epfl.ch/publications/uploadedFiles/TRO2316022.pdf
[5]: http://www.coppeliarobotics.com
[6]: https://pypi.org/project/pybullet
[7]: https://github.com/robotology-playground/pybullet-robot-envs
[8]: https://github.com/robotology/gym-ignition

**Comment by teresalves** [comment]
Thank you so much for all the information! 
This will certainly be very helpful for my research.

Kind Regards,
Teresa

**Comment by plinioMoreno** [comment]
Hi @pattacini and @xEnVrE,

Many thanks for the detailed answer. @teresalves wants to use the tactile sensors on the iCub hand as main observation to infer which kind of object is (we will start with simple objects like cylinders, parallelograms, spheres), based on the hand pose and the tactile information. In addition, the objects at our initial tests can be attached to the ground.

Our main question is if considering the developments by @xEnVrE , we can move the hand to a set of predefined points, close the hand until the tactile sensors reach a predefined threshold and store the data samples.

Looking forward to your answer.

**Comment by pattacini** [comment]
Hi @plinioMoreno 

I would say definitely yes given the work @xEnVrE did in this respect, but @xEnVrE can say the last word on it.

Let me only add to the discussion that you may find this article interesting: https://ieeexplore.ieee.org/document/7803305.

cc @lornat75

**Comment by xEnVrE** [comment]
Hi @plinioMoreno 

I would also say that the experiment is doable in a simulated setting like the one I was working on. 

However, let me add some considerations:

> infer which kind of object is (we will start with simple objects like cylinders, parallelograms, spheres), based on the hand pose and the tactile information

while the hand pose is available within the simulated environment, the tactile information (implemented in one of the plugins that I added) only consists of a list of the 3D points on the fingertips that are in contact with the environment. Hence, this information can be used to detect contacts and find their locations. No information on the exchanged forces is provided.

>  the objects at our initial tests can be attached to the ground

this hypothesis can be easily enforced within Gazebo by using a dummy joint between the object and the ground. This will help you a lot since some objects tend to move in the simulation environment, due to the exchanged contacts with the environment, even if no other external agents move them. Of course, in case you need unconstrained objects in future steps of the project, you can alleviate the phenomenon by playing with the weight of the object and/or the friction between the object and the ground.

> we can move the hand to a set of predefined points

this is already working but some effort might be needed to ensure that the desired hand pose is actually reached - especially if you need enough precision to ensure that the finger will touch the object. If not, this would require to tune a bit the low-level PID gains of the simulated robot.

>  close the hand until the tactile sensors reach a predefined threshold 

as I said before, right now the plugins implementing the tactile information only provides contact locations. However, you can either stop the motion as soon as a contact is detected or implement also some kind of force sensing within the plugin. In the second case, I can provide further explanations on the code of the plugin.

Please let me know in case you need further information.

**Comment by plinioMoreno** [comment]
Many thanks @pattacini and @xEnVrE ! @teresalves will set up this and we will let you know.
